### PR TITLE
feat: filter chunk assets

### PR DESCRIPTION
**Pull request by cahnory**
## Summary

In development, when doing SSR with HMR you don't want the hmr related files to be linked by the served html. The `filter` option in addition to the new `info` property added to chunk assets make it possible.

I favored this approach rather than an _"excludeHmrFiles"_ type option as it could serve other purposes and remain relevant even if hmr is abandoned in the future.

## Test plan

Using webpack with hmr enabled, instanciate your `ChunkExtractor` like :

```js
import { ChunkExtractor } from '@loadable/server'

const extractor = new ChunkExtractor({
  statsFile: '../dist/loadable-stats.json',
  filter: ({ info }) => !info.hotModuleReplacement,
});
```

All files relative to hmr (suffixed with _"-wps-hmr"_ by default) should be excluded from the results of `getScriptTags`, `getScriptElements`,…

### CODE REVIEW

**Comment by theKashey** [comment]
Fixes #652 ?

File: packages/server/src/ChunkExtractor.js [hunk 1]
```diff
@@ -187,8 +188,13 @@ class ChunkExtractor {
     this.outputPath = outputPath || this.stats.outputPath
     this.statsFile = statsFile
     this.entrypoints = Array.isArray(entrypoints) ? entrypoints : [entrypoints]
+    this.filter = filter || Boolean
```
Comment (theKashey) [inline packages/server/src/ChunkExtractor.js:191]: Let's use a less generic name on `this` scope. For example `filterChunk`

**Comment by cahnory** [comment]
> Fixes #652 ?

I don't think so.

I didn't want to mess with _loadable-stats.json_ as my intent was only to prevent serving hmr chunks but the stats could potentially serve other purposes. 
So this PR allows `ChunkExtractor` to apply a filter on the assets it returns via its getter methods while #652 would requires the webpack plugin to apply a filter on the chunks it add to the _loadable-stats.json_ file.

File: packages/server/src/ChunkExtractor.js (hunk 1 above)
Comment (cahnory) [reply packages/server/src/ChunkExtractor.js]: You're right, I made the change.

Comment (cahnory) [reply packages/server/src/ChunkExtractor.js]: I also rename the option as it makes its purpose more explicit.

**Comment by cahnory** [comment]
Don't know why the ci failed on the last commit.

**Comment by cahnory** [comment]
@theKashey if you prefer that the filter occurs in the webpack plugin (exclude chunks from stats file) let me know.

File: packages/server/src/ChunkExtractor.test.js [hunk 2]
```diff
@@ -299,6 +307,55 @@ describe('ChunkExtrator', () => {
...
+            async={true}
+            data-chunk="letters-A"
+            nonce="testnonce"
+            src="/dist/node/7415d90-letters-A-wps-hmr.js"
+          />,
+        ]
+      `)
+    })
+
+    it.only('should use filter from options', () => {
```
Comment (theKashey) [inline packages/server/src/ChunkExtractor.test.js:320]: `.only` should be removed 😅

**Review by theKashey — CHANGES_REQUESTED** [review]

File: packages/server/src/ChunkExtractor.test.js (hunk 2 above)
Comment (cahnory) [reply packages/server/src/ChunkExtractor.test.js:320]: Woopsy ^^

**Comment by cahnory** [comment]
@theKashey I'm having a problem with _stats.json_.

 When generated, it does not contain hmr chunks (obviously) and the `assets.*.info` property is always empty (but maybe the `hotModuleReplacement` property is only set when `true`).
I can manipulate stats object in my test to match my needs but I think it's not the way I'm supposed to do this.

**Comment by theKashey** [comment]
Yeah, probably you should take the original stats (as it's now generated from a real project) and add some "extra" fields in runtime to test, or filter some "normal" chunks, as it does not have test `hot` case, only the filtering

**Comment by cahnory** [comment]
> or filter some "normal" chunks, as it does not have test `hot` case, only the filtering

I had my initial usecase so much on my mind that I didn't think about it!

**Comment by theKashey** [comment]
👀 .....

**Comment by cahnory** [comment]
😅 I don't know what's wrong. I run the commands `ci`, `test` and even `build` locally without any problem. It seems to me that it has something to do with a deployment step that I don't master. In netlify logs I just see :
> failed Building production JavaScript and CSS bundles

Any idea ?

**Comment by theKashey** [comment]
@cahnory - it's just dead a little. Strangely there is no travis here, and these tests are the only really important thing.

**Comment by eliseumds** [comment]
Anyway I can help here?

**Comment by theKashey** [comment]
The real problem here is a "new" undocumented feature, which implementation is still on the user.
If HMR temporal files should not be a part of webpack builds - then they should not. Full stop and no extra conditions.

**Comment by cahnory** [comment]
> The real problem here is a "new" undocumented feature, which implementation is still on the user.
> If HMR temporal files should not be a part of webpack builds - then they should not. Full stop and no extra conditions.

This is not what I understood from our previous exchanges. In this case, if I understand correctly, it seems unsolvable and maybe we should close this PR.

**Comment by theKashey** [comment]
> This is not what I understood from our previous exchanges.

Decisions change in time. That's ok.

>  if I understand correctly, it seems unsolvable

The internal changes are legit and I would like to have them. I am just thinking that we should, at least for now, put `filter: ({ info }) => !info.hotModuleReplacement,` __inside__ loadable logic, keep external API intact as and I am not sure is there any other use case for `filter`. If you have one - feel free to change my mind.

The _lazy maintainer rule_ is simple - if you can __not expose__ something - do not, or you will have to maintain it forever (and you are lazy by definition).

**Comment by cahnory** [comment]
I understand, no problem. Just that by doing this way this PR should be closed in favor of a new one which maybe just add a cookbook recipe to the doc.

The way I do it right now is just by filtering `extractor.getXElements` results with this function:
```js
const filterHmr = (list) =>
  list.filter(
    (element) =>
      !element.props.src?.match(/wps-hmr\.[^.\\/]+$/) &&
      !element.props.href?.match(/wps-hmr\.[^.\\/]+$/),
  );
```

I just wasn't satisfied with having to check the file names rather than their stats (and extra loops but they are negligible).

**Comment by theKashey** [comment]
That's why it's still good to have the majority of this PR built-in into loadable with the correct behavior working out of the box

**Comment by cahnory** [comment]
> That's why it's still good to have the majority of this PR built-in into loadable with the correct behavior working out of the box

Oh I see, I completely misunderstood what you were saying. Sorry.
 I'll try to find a moment to make the changes.

**Pull request closed without merging**